### PR TITLE
TTT: Player and spectator chat channels

### PR DESCRIPTION
**Pull request by Kefta**
No more shitty redirects, no more tag ghosting. Uses a series of checks in PlayerCanSeePlayersChat (thanks Robot) to direct who can view chat. Tested extensively for the past two days with multiple people.

### CODE REVIEW

**Comment by comedinha** [comment]
It is possible traitors talk only with traitors and traitors talk with innocent?

**Comment by Kefta** [comment]
Yes. Everyone can see normal chat, but only traitors/detectives and see their respective team chats. Innocent team chat will just redirect to the main chat

File: garrysmod/gamemodes/terrortown/gamemode/gamemsg.lua [hunk 1]
```diff
@@ -91,54 +92,59 @@ end
...
+      elseif team_only and not team and ply:IsSpecial() then
+	     RoleChatMsg(ply, ply:GetRole(), text)
+		 return ""
       end
-
-      table.insert( filtered, 1, "[MUMBLED]")
-      return table.concat(filtered, " ")
    end
-
+
```
Comment (zerfgog) [inline garrysmod/gamemodes/terrortown/gamemode/gamemsg.lua]: hnng

Comment (Kefta) [inline garrysmod/gamemodes/terrortown/gamemode/gamemsg.lua]: Sorry there, pal

Comment (zerfgog) [inline garrysmod/gamemodes/terrortown/gamemode/gamemsg.lua]: you ought to fix that, pal
(ง'̀-'́)ง

Comment (Kefta) [inline garrysmod/gamemodes/terrortown/gamemode/gamemsg.lua]: Submit your own PR; I'm not adding another commit, pal :last_quarter_moon_with_face:

**Comment by comedinha** [comment]
When innocents say on the chat show (TEAM) message

**Comment by comedinha** [comment]
This make the player has proven

**Comment by Kefta** [comment]
Can you check if that chatbox overrides the OnPlayerChat hook? I made it so that it doesn't on the default one

**Comment by meepdarknessmeep** [comment]
that might be because of your overridden chatbox addon not properly handling it

**Comment by Kefta** [comment]
I rather not create a hook to work around shitty chatboxes, but I might have to.

**Comment by comedinha** [comment]
net.Receive("atlaschat.chatText", function(bits)
	local text = net.ReadString()
	local player = net.ReadEntity()
	local team = util.tobool(net.ReadBit())
	local dead = IsValid(player) and !player:Alive() or false

	hook.Run("OnPlayerChat", player, text, team, dead)
end)

**Comment by Kefta** [comment]
What's the net.Send code?

**Comment by meepdarknessmeep** [comment]
local player = net.ReadEntity()

pls no

**Comment by Kefta** [comment]
Yeah, I've had to help people a lot with Atlaschat on FP. Not the best

**Comment by Chewgum** [comment]
What is the issue with that @meepdarknessmeep ?

**Comment by comedinha** [comment]
I created a ticket with the creator of the Atlas chat

**Comment by meepdarknessmeep** [comment]
net.WriteString("!ban ^ 0 you are a BITCH")
net.WriteEntity(admin)
net.WriteBit(0)

**Comment by Chewgum** [comment]
What?

**Comment by wiox** [comment]
Meep's dumb.

**Comment by meepdarknessmeep** [comment]
nevermind it's a clientside hook /suicide

**Comment by Chewgum** [comment]
My chatbox does not override anything and it uses whatever the OnPlayerChat hook passes onto chat.AddText. I also run the "PlayerSay" hook for the text and the "PlayerCanSeePlayersChat" hook when I build my filter. Either comedinha is using some very old version or he/she has some other addon that is messing with chat. I did have to update my chat with the removal of "LANG.Msg("spec_teamchat_hint")" though, but that shouldn't cause this issue..

net.Receive("atlaschat.chat", function(bits, player)
	if (player.nextMessageAtlas > CurTime()) then return end
	
	player.nextMessageAtlas = CurTime() +0.25
	
	local text = net.ReadString()
	local team = util.tobool(net.ReadBit())
	
	text = string.sub(text, 0, 127)
	
	local newText = hook.Run("PlayerSay", player, text, team, !player:Alive())
	
	-- A workaround for dumb coders that return a boolean in the PlayerSay hook.
	if (isstring(newText)) then
		text = newText
	elseif (not newText) then
		text = ""
	end
	
	if (text != "") then
		if (game.IsDedicated()) then
			ServerLog(player:Nick() .. ": " .. text .. "\n")
		end
		
		local filter = {}
		local players = util.GetPlayers()
		
		for i = 1, #players do
			local target = players[i]
			
			if (IsValid(target)) then
				local canSee = hook.Run("PlayerCanSeePlayersChat", text, team, target, player)
				
				if (canSee or target == player) then
					table.insert(filter, target)
				end
			end
		end
		
		net.Start("atlaschat.chatText")
			net.WriteString(text)
			net.WriteEntity(player)
			net.WriteBit(team)
		net.Send(filter)
	end
end)

net.Receive("atlaschat.chatText", function(bits)
	local text = net.ReadString()
	local player = net.ReadEntity()
	local team = util.tobool(net.ReadBit())
	local dead = IsValid(player) and !player:Alive() or false

	hook.Run("OnPlayerChat", player, text, team, dead)
end)

**Comment by comedinha** [comment]
https://github.com/jkralicky/Custom-ULX-Commands
I found the server problem

**Comment by Kefta** [comment]
https://github.com/jkralicky/Custom-ULX-Commands/blob/bc0ede4ced0b7ea6ebff273828a51954d7a4d057/CustomCommands/lua/ulx/modules/sh/cc_chattags.lua#L49

Yeah, let's just copy the baseclass function in a custom hook 👍

**Comment by CSchulz** [comment]
Pressing y results in "[ERROR: Translation of spec_teamchat_hint not found]".

**Comment by Kefta** [comment]
@CSchulz Cannot replicate. Can you please download the terrortown gamemode from the master branch and report back if the error is still occuring?

**Comment by CSchulz** [comment]
So Translations aren't synchronized from the server to the client?

**Comment by Kefta** [comment]
@CSchulz No, I removed that translation since it isn't used anymore with this change. That's why I asked you to redownload the gamemode since it seems you have outdated language files.

**Comment by wiox** [comment]
Could be the custom chatbox mentioned above (or another).

**Comment by CSchulz** [comment]
Alright so nevermind, I hoped the files are synchronized. So I need to wait til the new client is available.

**Comment by Chewgum** [comment]
The current version of my chatbox is still using "spec_teamchat_hint" since the version of TTT on gmod still has that. If you're using my chatbox (atlaschat) CShulz, you'll get that errror.

**Comment by doctorluk** [comment]
Can confirm the problem @comedinha has reported.

In TTT, when you're innocent, you can use the Team chat, which everyone can see, but Traitors cannot use. This automatically results in this player being prooven innocent, which essentially breaks the game.

**Comment by Kefta** [comment]
@doctorluk We have already traced the problem down to outdated addons. Remove them

**Comment by doctorluk** [comment]
I do not use the mentioned addon above, I'll try to track it down in the ULX forums. Thanks for the hint!
